src/gui/tagfiltermodel.cpp

```diff
@@ -0,0 +1,326 @@
+{
+    if (!index.isValid()) return QString();
+    return static_cast<TagModelItem *>(index.internalPointer())->tag();
+}
+
+QModelIndex TagFilterModel::index(const TagModelItem *item) const
+{
+    if (!item)
+        return QModelIndex();
+    const int row = m_tagRows.indexOf(item->tag());
+    if (row < 0)
+        return QModelIndex();
+    return index(row, 0, QModelIndex());
+}
+
+void TagFilterModel::tagAdded(const QString &tag)
+{
+    const int row = m_tagRows.count();
+    beginInsertRows(QModelIndex(), row, row);
+    addItemToModel(new TagModelItem(tag, tag));
+    endInsertRows();
+}
+
+void TagFilterModel::tagRemoved(const QString &tag)
+{
+    TagModelItem *item = findItem(tag);
+    if (item) {
+        QModelIndex i = index(item);
+        beginRemoveRows(i.parent(), i.row(), i.row());
+        removeItemFromModel(item);
+        endRemoveRows();
+    }
+}
+
+void TagFilterModel::torrentTagAdded(BitTorrent::TorrentHandle *const torrent, const QString &tag)
+{
+    if (torrent->tags().count() == 1)
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->increaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentTagRemoved(BitTorrent::TorrentHandle* const torrent, const QString &tag)
+{
+    Q_ASSERT(torrent->tags().count() >= 0);
+    if (torrent->tags().count() == 0)
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    TagModelItem *item = findItem(tag);
+    Q_ASSERT(item);
+
+    item->decreaseTorrentsCount();
+    const QModelIndex i = index(item);
+    emit dataChanged(i, i);
+}
+
+void TagFilterModel::torrentAdded(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->increaseTorrentsCount();
+
+    const QVector<TagModelItem *> items = findItems(torrent->tags());
+    if (items.isEmpty())
+        findItem(getSpecialUntaggedTag())->increaseTorrentsCount();
+
+    foreach (TagModelItem *item, items)
+        item->increaseTorrentsCount();
+}
+
+void TagFilterModel::torrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent)
+{
+    findItem(getSpecialAllTag())->decreaseTorrentsCount();
+
+    if (torrent->tags().isEmpty())
+        findItem(getSpecialUntaggedTag())->decreaseTorrentsCount();
+
+    foreach (TagModelItem *item, findItems(torrent->tags()))
+        item->decreaseTorrentsCount();
+}
+
+QString TagFilterModel::getSpecialAllTag()
+{
+    static const QString *const ALL_TAG = new QString(" ");
+    Q_ASSERT(!BitTorrent::Session::isValidTag(*ALL_TAG));
+    return *ALL_TAG;
+}
+
+QString TagFilterModel::getSpecialUntaggedTag()
+{
+    static const QString *const UNTAGGED_TAG = new QString("  ");
+    Q_ASSERT(!BitTorrent::Session::isValidTag(*UNTAGGED_TAG));
+    return *UNTAGGED_TAG;
+}
+
+void TagFilterModel::populate()
+{
+    using Torrent = BitTorrent::TorrentHandle;
```

Comment:
IIRC, it was like this in the code I copied over from the CategoryFilterModel. If you don't like it, I'm happy to change it.